src/lib.rs

```diff
@@ -76,6 +114,176 @@ pub fn encode_vec_with_sentinel(source: &[u8], sentinel: u8) -> Vec<u8> {
+    /// decoding to complete
+    pub fn feed(&mut self, data: u8) -> Result<Option<usize>, usize> {
+        use DecoderState::*;
+
+        let (ret, state) = match (&self.state, data) {
+            // Currently Idle, received a terminator, ignore, stay idle
+            (Idle, 0x00) => (Ok(None), Idle),
+
+            // Currently Idle, received a byte indicating the
+            // next 255 bytes have no zeroes, so we will have 254 unmodified
+            // data bytes, then an overhead byte
+            (Idle, 0xFF) => (Ok(None), GrabChain(0xFE)),
+
+            // Currently Idle, received a byte indicating there will be a
+            // zero that must be modified in the next 1..=254 bytes
+            (Idle, n)    => (Ok(None), Grab(n - 1)),
+
+            // We have reached the end of a data run indicated by an overhead
+            // byte, AND we have recieved the message terminator. This was a
+            // well framed message!
+            (Grab(0), 0x00) => (Ok(Some(self.dest_idx)), ErrOrComplete),
+
+            // We have reached the end of a data run indicated by an overhead
+            // byte, and the next segment of 254 bytes will have no modified
+            // sentinel bytes
+            (Grab(0), 0xFF) => {
+                add(self.dest, self.dest_idx, 0u8).map_err(|_| self.dest_idx)?;
+                self.dest_idx += 1usize;
+                (Ok(None), GrabChain(0xFE))
+            },
+
+            // We have reached the end of a data run indicated by an overhead
+            // byte, and we will treat this byte as a modified sentinel byte.
+            // place the sentinel byte in the output, and begin processing the
+            // next non-sentinel sequence
+            (Grab(0), n) => {
+                add(self.dest, self.dest_idx, 0u8).map_err(|_| self.dest_idx)?;
+                self.dest_idx += 1usize;
+                (Ok(None), Grab(n - 1))
+            },
+
+            // We were not expecting the sequence to terminate, but here we are.
+            // Report an error due to early terminated message
+            (Grab(_), 0) => {
+                (Err(self.dest_idx), ErrOrComplete)
+            }
+
+            // We have not yet reached the end of a data run, decrement the run
+            // counter, and place the byte into the decoded output
+            (Grab(i), n) =>  {
+                add(self.dest, self.dest_idx, n).map_err(|_| self.dest_idx)?;
+                self.dest_idx += 1;
+                (Ok(None), Grab(i - 1))
+            },
+
+            // We have reached the end of a data run, and we expect a data header
+            // here. Since 0x00 is not a valid overhead byte, this is an error case.
+            // This may non *NEED* to be an error, but Example 9 on
+            // https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing
+            // indicates that this is not correct behavior
+            (GrabChain(0), 0x00) => {
```

Comment:
Hey @awelkie, let me know what you think about this state transition. Example 9 of https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing shows that they expect an extra overhead byte after a 254 byte "GrabChain", even if it signifies that the next byte is a zero.

I think this is superfluous, but maybe I am missing something.

Once we have a decision, I can write a test that works one way and fails the other, so the ambiguity is resolved.

Reply:
Oops, I misunderstood the test set. There is a trailing zero that causes this to happen. I'll correct the code.